### ISSUE TITLE
Live activity

### ISSUE THREAD

**Issue by miguelpruivo** [JohanDegraeve/xdripswift#446]
Can we have [live activity](https://developer.apple.com/documentation/activitykit/displaying-live-data-with-live-activities) to have the BG always on the screen like we have for example in other apps? That adds a nice convenience to the app. People with dynamic island also would have the BG always in it.

**Comment by julesjwoods** [comment]
Yeah, although xDrip is all round a significantly better app, this is a game changer of a feature used in GlucoseDirect.

**Comment by miguelruivo** [comment]
> Yeah, although xDrip is all round a significantly better app, this is a game changer of a feature used in GlucoseDirect.

Yeah, that's where I saw it in the first place, but reverted back to xDrip because of the alarms customization that GlucoseDirect doesn't have.

**Comment by miguelpruivo** [comment]
Yeah, that's where I saw it in the first place, but reverted back to xDrip because of the alarms customization that GlucoseDirect doesn't have

**Comment by miguelpruivo** [comment]
> Yeah, although xDrip is all round a significantly better app, this is a game changer of a feature used in GlucoseDirect.

Yeah, that's where I saw it in the first place, but reverted back to xDrip because of the alarms customization that GlucoseDirect doesn't have

**Comment by paulplant** [comment]
OK... on the list

**Comment by afshopify** [comment]
@paulplant it looks like this is disabled for anyone in follower mode. Wondering if we can enable it for those in follower mode?

**Comment by aug0211** [comment]
@paulplant it looks like this is disabled for anyone in follower mode. Wondering if we can enable it for those in follower mode?

**Comment by paulplant** [comment]
It's disabled for follower modes that are using the background audio keep-alive because it won't allow them to update.

I added a check to allow them to be enabled for follower modes when using an external heartbeat to wake the app up.

**Comment by aug0211** [comment]
Oh, background audio won’t allow it up update? I hadn’t realized that.

If the push alerts work, would live activity also work?

**Comment by aug0211** [comment]
I haven't done much with live activities before, maybe a simple question - but does the developer mode toggle that works for widget updates also allow live activity to update on the increased dev cycle (perhaps with xDrip set to use the aggressive audio backgrounder)?

**Comment by paulplant** [comment]
> Oh, background audio won’t allow it up update? I hadn’t realized that.
Live activities cannot update for an app that is in the background. The silent audio works to partially wake the app up enough to quickly process follower downloads and trigger alarms, but the app is still officially in a background state.

> If the push alerts work, would live activity also work?
I'm not sure what you mean? Do you mean APNs? How is that related?

**Comment by paulplant** [comment]
> I haven't done much with live activities before, maybe a simple question - but does the developer mode toggle that works for widget updates also allow live activity to update on the increased dev cycle (perhaps with xDrip set to use the aggressive audio backgrounder)?

I don't follow your question, sorry.

Setting WidgetKit in Developer mode allows you to basically ignore refresh constraints but this isn't really noticeable anyway for Live Activities. They are allowed very frequent updates. 

It's only really noticeable for standard iOS widgets (but only just) and definitely for WatchOS widgets (complications).

But it's not related to anything to do with background keep-alive.

Can you explain what it is you're trying to figure out?

**Comment by aug0211** [comment]
Sure, sorry. Like I said, I have not done much with live activity before so my questions may not be going the right direction 😅

What I’m going after is live activities in follower mode without a heartbeat. Eg, a caregiver following a kiddo, and getting live activities on the caregiver device.

I was hoping that the same way we can get semi-reliable updates for widgets and for push alerts that we could get live activities updating semi-regularly in follower mode.

I thought perhaps the same triggers used for widget refreshes or push notifications could be useful for live activities. I also wasn’t sure if live activities leverage the same iOS refresh cadence policies as widgets, and if so, if the developer widget refresh toggle might help.

**Comment by paulplant** [comment]
Live activities *won't work* in follower mode without a heartbeat. If they could, we wouldn't have purposefully disabled them.

Although all are spawned from the WidgetKit framework, the constraints are very different. Live activities are designed for frequent updates but are not supposed to be permanently used in the way we are using them. The developer toggle won't make any difference as it is for a different thing than you're talking about.

It would be possible to wake up the app and update the live activities in the background if we were pushing values to the app from a web service via APNs, but that isn't the case with xDrip4iOS.

An easy way to get around this would be to use a RileyLink or EmaLink and carry it around in your pocket and use it as an external heartbeat for xDrip4iOS. I've tested this and it works perfectly.

**Comment by aug0211** [comment]
Thanks Paul for helping me learn some. I've also dug up some documentation on live activities to learn more (eg, I didn't know they could run for up to 8 hours, that seems promising somehow).

I also took your advice and went ahead and ordered an OrangeLink Pro for now. Hopefully there's a creative option that emerges in time.

**Comment by aug0211** [comment]
@paulplant, curious if you know of any documentation (or maybe a patch is needed?) to unlock using an OrangeLink as my heartbeat for a Follower (parent/caregiver) device. Or if you could point me in the direction for how you did this when you tested with it.

I may have missed this in published docs, sorry if so!

**Comment by aug0211** [comment]
I noticed something odd today. BG deltas are always reporting as +0 in widget and live actions views. In the app, it's showing the appropriate BG delta between readings (+5, -11, etc). This happened after deploying 5.2 over top of master.

This is with the heartbeat approach being used with follower mode, and no code customizations. I'm seeing this in master, staging 5.2, and develop branches.

If you'd like me to try anything for testing or to help debug, I'm happy to.

![image](https://github.com/JohanDegraeve/xdripswift/assets/659845/3a4f813b-a324-49fd-bd80-94b0949dbdc0)![image](https://github.com/JohanDegraeve/xdripswift/assets/659845/360f33b0-d111-4d68-92d3-677ae91cb46a)

**Comment by paulplant** [comment]
Looking at your chart, it looks like you have duplicated readings. Please tap the BG Readings button at the top left of the Home Screen (2nd icon) and show us if this is the case.

Are you sure your master device is not uploading to Nightscout from two apps at the same time?

Actually, just checked the Nightscout URL (you can double tap this to hide it for 10 seconds to take screenshots)... you have duplicate readings being uploaded every 5 minutes:

<img width="580" alt="image" src="https://github.com/JohanDegraeve/xdripswift/assets/37302780/480353ae-4941-495e-932d-4cc06df22846">

**Comment by aug0211** [comment]
Yes 100% this was it! I’m sorry I discovered this and tried to delete my comment before you spent any time looking at it 🤦‍♂️

The issue was I had NS enabled on the master device, so it was uploading in addition to iAPS. I turned off NS upload and this resolved it.

I have seen *some* occasions where BG delta reports oddly as +0 when it’s actually a -1 or +1, but haven’t gathered enough info around that yet. Figured I share since we’re talking about it here, but I think it’s unrelated/a different scenario. If I recall correctly, the live activity and widget had this issue but I think maybe the main view did not. I’ll watch and gather more info!

Thanks again Paul!

**Comment by aug0211** [comment]
Ok, I got a bit more info. We had pretty steady BG this AM with a handful of small deltas like -1, and that seemed to trigger the inaccurate delta of +0 being reported.

Let me know if you'd like me to create another issue for this one. I've noticed it in Live Activity displays/Widget, iOS Standby display, and the app main display mode.

The issue is xDrip4iO5 displays a BG delta of +0 across all surfaces when it is in fact not +0.

It is reproducible and consistent in all cases of -1s I've been able to see (maybe 3-5 so far today and 5-10 overall). Other deltas seem accurate (not 100% confident on +1, haven't been able to spot that yet, but others like +5 or -8 etc look fine).

I confirmed only one device is uploading to NS.

Included screen shots below showing the +0 reported by xDrip4iO5, whereas Nightscout correctly shows the delta of -1.![image](https://github.com/JohanDegraeve/xdripswift/assets/659845/a27c3bb2-a9e1-4c74-ab40-1e73439548b7)![image](https://github.com/JohanDegraeve/xdripswift/assets/659845/a3d05ac4-20de-4ad2-8a91-645fe1bd316f)

**Comment by paulplant** [comment]
But if your screenshot, you can clearly see that the latest reading is lower than the previous one... so technically it shouldn't be +0...

<img width="84" alt="image" src="https://github.com/JohanDegraeve/xdripswift/assets/37302780/62ff712f-cca6-4cdd-8794-4dea1a289c78">

I think this is probably just a rounding issue for readings that are right on the limit... I wouldn't worry about it too much...

**Comment by paulplant** [comment]
Can you click the BG Readings view (the drop icon on the Home Screen), screenshot it, then click the two readings that are from the above screenshots and screenshot the details screen for each one.

**Comment by aug0211** [comment]
Here's a +1 delta example. Maybe this should be a separate issue, as it doesn't seem to be limited to only Live Activities. Let me know if you want me to create a different issue.

Screen shots below showing main Home Screen/widget/Live Activity from xDrip4iO5 showing +0, alongside Nightscout (via browser) and Dexcom (via LoopFollow) showing the actual +1.

Poking around a bit in the code to see if I can pinpoint this but no luck yet. I see some logic for handling corner cases around 0 but nothing jumping out yet.

Here are the two most recent readings from MongoDB database as well. All looks in order to me there.

9:50 AM
```
{"_id":"82BBE7DB-5670-4A87-85ED-3C702E5D0279","date":{"$numberDouble":"1713883839850.0"},"direction":"Flat","dateString":"2024-04-23T14:50:39.850Z","sgv":{"$numberInt":"103"},"glucose":{"$numberInt":"103"},"type":"sgv","activationDate":"2024-04-19T12:02:21.562Z","sessionStartDate":"2024-04-19T12:27:21.562Z","unfiltered":{"$numberInt":"103"},"utcOffset":{"$numberInt":"0"},"sysTime":"2024-04-23T14:50:39.850Z"}
```

9:45 AM
```
{"_id":"AE0C990D-536F-44B7-A082-E653EE24B2AC","dateString":"2024-04-23T14:45:38.574Z","unfiltered":{"$numberInt":"102"},"sgv":{"$numberInt":"102"},"date":{"$numberDouble":"1713883538573.0"},"type":"sgv","direction":"Flat","glucose":{"$numberInt":"102"},"activationDate":"2024-04-19T12:02:21.562Z","sessionStartDate":"2024-04-19T12:27:21.562Z","utcOffset":{"$numberInt":"0"},"sysTime":"2024-04-23T14:45:38.574Z"}
```
![image](https://github.com/JohanDegraeve/xdripswift/assets/659845/100622ba-f145-4270-b661-3133b9025adf)![image](https://github.com/JohanDegraeve/xdripswift/assets/659845/d6a790ac-c7f8-47ca-ac0c-de42898707b1)![image](https://github.com/JohanDegraeve/xdripswift/assets/659845/998ab3b9-9250-4012-9e10-899909af2bfe)![image](https://github.com/JohanDegraeve/xdripswift/assets/659845/56100641-75d1-43fb-a4c6-040a15b926db)![image](https://github.com/JohanDegraeve/xdripswift/assets/659845/dcc4ef13-d4fc-4749-98c5-a32cbec12a8d)

**Comment by aug0211** [comment]
![image](https://github.com/JohanDegraeve/xdripswift/assets/659845/eccfe4ef-3d93-45e0-939b-5c64d6e5babd)![image](https://github.com/JohanDegraeve/xdripswift/assets/659845/4b2c8e93-88c0-41e7-bf2b-26bdb14ed2bf)![image](https://github.com/JohanDegraeve/xdripswift/assets/659845/3ae31a93-0a60-4cf5-aa54-99003c3fe8bb)

**Comment by aug0211** [comment]
Some other interesting info I thought of that might be helpful:

- This is on a device in **Follower** mode
- I'm pulling data from NS, originally sent from iAPS on the main phone
- The sensor is a G7
- The main phone also has xDrip4iO5 running (same version, main)
- xDrip4iO5 in the main phone is not uploading readings to NS (iAPS is)
- **The main device with the same version of xDrip4iO5 does not seem show this issue, only the following device does** (validated just for one reading just now, anyway - where my following device showed the issue but the main device did not).

**Comment by paulplant** [comment]
Nice info... thanks!!

**Comment by aug0211** [comment]
I'm curious, do we know if I'm the only one who has seen this? Wondering if others in Follow mode have noticed it.

**Comment by paulplant** [comment]
I'm running two versions in parallel. My master app uploading to Nightscout, and a follower downloading the same values from Nightscout... and the deltas are always (so far) correct and aligned... I've been running it like this for 3 days and I've never noticed a difference.

Do you think you could share your Nightscout URL with so that I can try it and try to figure out what is going on? You can send it by e-mail to xdrip@proximus.be

**Comment by aug0211** [comment]
Interesting! Just sent it.

**Comment by aug0211** [comment]
You're looking at the +1 and -1 cases right? That's where I see false +0s.

**Comment by aug0211** [comment]
I think I had some that were accurate when I uploaded from xDrip rather than from iAPS. For a bit I thought they were all correct - but I did still manage to spot a few that were off.

**Comment by aug0211** [comment]
@JohanDegraeve heads up, updated to latest release today (on master branch). I'm using the Heartbeat method you helped me with (thank you again, total game changer).

**I've lost widgets, live activities, and watch complications on my device since updating.**

I've tried flipping to master then back to follower, turning off/on live activities in iOS settings, turning off/on live activities in xDrip, re-installing app (on phone and watch), removing and re-adding my heartbeat, etc; but no luck.

It seems like xDrip is no longer recognizing me as "allowed" to use widgets/live activities. I recall there are some checks to prevent invalid scenarios (follower without a heartbeat). Maybe I'm somehow caught up in that check.

Master or Follower: Follower
Follower Data Source: Nightscout
Background Keep-alive: Heartbeat 🩶

Live Activities: Always
Live Activity Size: Large

Let me know if other info would be helpful to provide.

**Comment by aug0211** [comment]
![IMG_3297](https://github.com/JohanDegraeve/xdripswift/assets/659845/eb84885c-24a5-46a2-a1d4-040ba94c4c63)
![IMG_3298](https://github.com/JohanDegraeve/xdripswift/assets/659845/6e45ca20-96ec-4600-891a-7a70debb279f)
![IMG_3299](https://github.com/JohanDegraeve/xdripswift/assets/659845/b627234a-acac-4f03-ad40-8b8a988e33f0)
![IMG_3300](https://github.com/JohanDegraeve/xdripswift/assets/659845/bed16ad2-d315-434b-a61a-12a4dd9f2dc8)
![IMG_3301](https://github.com/JohanDegraeve/xdripswift/assets/659845/7ac71e68-2636-47cf-a625-b651a51ff5ad)
![incoming-5653BF7A-8805-4F89-A932-3DBC76134037](https://github.com/JohanDegraeve/xdripswift/assets/659845/5d463aa5-9048-4032-80bd-cc9dbd0c8fbf)

**Comment by paulplant** [comment]
Change the schema to xDrip Widget and rebuild.Paul PlantOn 5 May 2024, at 15:15, Auggie ***@***.***> wrote:﻿
Intersting, tapping the dynamic island DOES launch the app.
It seems like it's in the dynamic island, but does not show any data. Dynamic island is just black, not showing any BG data.

**Comment by aug0211** [comment]
- Commit d290482 - issue still present (deploying from xdrip scheme)
- Commit a215d7a - issue still present (deploying from xdrip scheme)
- Commit a3ba279 - issue still present (deploying from xdrip scheme)
- **_Commit 29200cf - this commit works_** (deploying from xdrip scheme)

Just saw above note - deploying latest commit from "xdrip" scheme breaks, and then re-deploying latest commit from "xDrip Widget Extension" fixes the issue.

I think a single deployment from the main "xdrip" scheme is probably desired, not sure though. Thanks for the help with this workaround.

**Comment by paulplant** [comment]
Thanks for this. I'll look into it 👍

**Comment by paulplant** [comment]
Can you try and **disable** the Address Sanitizer function from the xDrip scheme -> Run -> Diagnostics?

And try and build the xDrip scheme and check that the widget extension is now installed correctly?

<img width="351" alt="image" src="https://github.com/JohanDegraeve/xdripswift/assets/37302780/e63dded3-b3ce-42a7-8088-bd9b139dcacc">

<img width="783" alt="image" src="https://github.com/JohanDegraeve/xdripswift/assets/37302780/d7cefaea-31d9-4f43-afc2-3d28fce76303">

**Comment by aug0211** [comment]
Absolutely. Will check. Kiddo birthday party now and didn't think to bring the laptop. Later tonight I expect!

**Comment by aug0211** [comment]
@JohanDegraeve this appears to have fixed it.

**Comment by paulplant** [comment]
OK thanks. So confirmed for #536
